s/start.go

```diff
@@ -45,13 +45,136 @@ func oneForOneRestart(
 		if newChild.Tag() == c.Worker {
 			eventNotifier.WorkerFailed(newChild.RuntimeName(), err)
 		}
-		return err
+		return c.Child{}, err
 	}
 
 	if newChild.Tag() == c.Worker {
 		eventNotifier.WorkerStarted(newChild.RuntimeName(), startTime)
 	}
-	return nil
+	return newChild, nil
+}
+
+func oneForOneRestartLoop(
+	eventNotifier EventNotifier,
+	supRuntimeName string,
+	supChildren map[string]c.Child,
+	supNotifyCh chan<- c.ChildNotification,
+	prevChild c.Child,
+) error {
+	for {
+		newChild, startErr := oneForOneRestart(
+			eventNotifier,
+			supRuntimeName,
+			supChildren,
+			supNotifyCh,
+			prevChild,
+		)
+		// if we don't get start errors, break the loop
+		if startErr == nil {
+			return nil
+		}
+
+		// otherwise, repeat until error treshold is met
+		// TODO: deal with restart error treshold here
+		prevChild = newChild
+	}
+}
+
+func handleChildError(
+	eventNotifier EventNotifier,
+	supRuntimeName string,
+	supChildren map[string]c.Child,
+	supNotifyCh chan c.ChildNotification,
+	prevChild c.Child,
+	prevChErr error,
+) error {
+	chSpec := prevChild.Spec()
+
+	eventNotifier.ProcessFailed(chSpec.Tag(), prevChild.RuntimeName(), prevChErr)
+
+	switch chSpec.GetRestart() {
+	case c.Permanent, c.Transient:
+		// On error scenarios, Permanent and Transient try as much as possible
+		// to restart the failing child
+		for {
+			return oneForOneRestartLoop(
+				eventNotifier,
+				supRuntimeName,
+				supChildren,
+				supNotifyCh,
+				prevChild,
+			)
+		}
+
+	default: /* Temporary */
+		// Temporary children can complete or fail, supervisor will not restart them
+		delete(supChildren, chSpec.Name())
+		return nil
+	}
+}
+
+func handleChildCompletion(
+	eventNotifier EventNotifier,
+	supRuntimeName string,
+	supChildren map[string]c.Child,
+	supNotifyCh chan c.ChildNotification,
+	prevChild c.Child,
+) error {
+	eventNotifier.WorkerCompleted(prevChild.RuntimeName())
+
+	chSpec := prevChild.Spec()
+
+	switch chSpec.GetRestart() {
+
+	case c.Transient, c.Temporary:
+		delete(supChildren, chSpec.Name())
+		// Do nothing
+		return nil
+	default: /* Permanent */
+		// On child completion, the supervisor still restart the child when the
+		// c.Restart is Permanent
+		for {
```

Comment:
Here it does make sense, the loop guarantees that the supervisor monitor won't stop to try start a child (Check docs for the values `Permanent` and `Transient` on [`c.Restart`](https://github.com/capatazlib/go-capataz/blob/d97d2fddf1b83cb31bb1ddbda4d675d20d6c930b/c/types.go#L31)), in this branch we are doing this forever, I've another PR to send your way with a error threshold tolerance check that would abort the restart procedure.

Reply:
But you are not checking anything. I just returns. There is no actual retry. What am I missing? When will `oneForOneRestartLoop()` ever be invoked more than once in that loop? There must be some language subtlety that I am completely missing.

Reply:
Argh... never mind my previous comment, I did not read the full code (only from the GH code snippet) and I thought this code was the `oneForOneRestart` instead of the `handleChilldCompletion` function. 

This for loop does not make sense either, is superfluous as well :man_facepalming: sorry, for the confusion here.